client/gutenberg/editor/api-middleware/index.js

```diff
@@ -34,26 +42,99 @@ export const pathRewriteMiddleware = ( options, next, siteSlug ) => {
 	return next( { ...options, path: wpcomPath } );
 };
 
+const wpcomRequest = method => {
+	if ( method !== 'GET' ) {
+		return wpcom.req.post.bind( wpcom.req );
+	}
+	return wpcom.req.get.bind( wpcom.req );
+};
+
 export const wpcomProxyMiddleware = parameters => {
 	// Make authenticated calls using the WordPress.com REST Proxy
 	// bypassing the apiFetch call that uses window.fetch.
 	// This intentionally breaks the middleware chain.
-	return new Promise( ( resolve, reject ) => {
-		const { body, data, ...options } = parameters;
+	const {
+		path,
+		body = {},
+		data,
+		method: rawMethod,
+		apiVersion,
+		apiNamespace = 'wp/v2',
+		transformResponse = identity,
+	} = parameters;
+
+	const method = rawMethod ? rawMethod.toUpperCase() : 'GET';
 
-		wpcomProxyRequest(
+	const payload = {};
+	if ( body.constructor.name === 'FormData' && body.has( 'file' ) ) {
+		// options.body is a FormData object which we need to convert to a plain old object
+		// Due to error below using Chrome and Safari using the wpcom proxy
+		// Error: Uncaught DOMException: Failed to execute 'postMessage' on 'Window': FormData object could not be cloned.
+		payload.data = toPairs( Array.from( body.entries() ) );
+		payload.formData = [ [ 'file', body.get( 'file' ) ] ];
+	} else if ( data || body ) {
+		payload.body = data || body;
+	}
+
+	return new Promise( ( resolve, reject ) => {
+		wpcomRequest( method )(
 			{
-				...options,
-				...( ( body || data ) && { body: body || data } ),
-				apiNamespace: 'wp/v2',
+				path,
+				method,
+				...( !! apiVersion && { apiVersion } ),
+				...( ! apiVersion && { apiNamespace } ),
+				...payload,
 			},
-			( error, bodyOrData ) => {
+			//error, data, headers
+			( error, dataResponse ) => {
 				if ( error ) {
 					return reject( error );
 				}
-
-				return resolve( bodyOrData );
+				return resolve( transformResponse( dataResponse ) );
 			}
 		);
 	} );
 };
+
+/**
+ * Transforms a v1.1 wpcom response to a wp/v2 response
+ * @param   {Object}  response the v1.1 oembed response
+ * @param   {String}  embedUrl the fallback embed url
+ * @returns {Object}  transformed response
+ */
+const transformOembedResponseFromWpcomToCore = ( response, embedUrl ) => {
```

Comment:
since we could get an `error` here we should probably rename this `maybeTransformOembedResponseFromWpcomToCore`

Reply:
Hah well that's one way to get me to refactor code :) updated in 445f074

Though I'll probably leave any polishing of error cases for follow ups